### PR TITLE
Suggested edits to "Get Started: Tune model parameters"

### PR DESCRIPTION
**Pull request by apreshill**
Mainly small edits. Bigger suggested edits involved adding some portions of text to:

1) flush out a bit more around those specific hyperparameters (please edit!)

1) the `dials::grid_regular()` bit (please edit!)

1) switched out the finalizing model bit at the end for finalizing a workflow (open to switching back, it did make it so you didn't have to repeat the formula over and over, and we introduced the `pull_workflow_*` functions in an earlier article).

### CODE REVIEW

File: content/start/tuning/index.Rmarkdown [hunk 1]
```diff
@@ -13,46 +13,68 @@ source(here::here("content/start/common.R"))
...
 
 - the complexity parameter (which we call `cost_complexity` in tidymodels) for the tree, and
 - the maximum `tree_depth`.
 
-Before we start the tuning process, we split our data into training and testing sets, just like when we trained the model with one default set of hyperparameters. We can use `strata = class` if we want our training and testing sets to be created using stratified sampling so that both have the same proportion of both kinds of segmentation.
+Tuning these hyperparameters often improves model performance because decision tree models are prone to [overfitting](http://www.feat.engineering/important-concepts.html#overfitting). This happens because single tree models tend to fit the training data _too well_ &mdash; so well, in fact, that they overlearn patterns and trends present in the training data that do not actually help when faced with new data to predict.
```
Comment (juliasilge) [inline content/start/tuning/index.Rmarkdown]: [](url)
```suggestion
Tuning these hyperparameters can improve model performance because decision tree models are prone to [overfitting](http://www.feat.engineering/important-concepts.html#overfitting). This happens because single tree models tend to fit the training data _too well_ &mdash; so well, in fact, that they over-learn patterns present in the training data that do not help when faced with new data to predict. 
```

File: content/start/tuning/index.Rmarkdown [hunk 2]
```diff
@@ -13,46 +13,68 @@ source(here::here("content/start/common.R"))
...
+Tuning these hyperparameters often improves model performance because decision tree models are prone to [overfitting](http://www.feat.engineering/important-concepts.html#overfitting). This happens because single tree models tend to fit the training data _too well_ &mdash; so well, in fact, that they overlearn patterns and trends present in the training data that do not actually help when faced with new data to predict. 
 
-```{r cell-split}
-library(tidymodels) 
+We will use hyperparameters to help our decision tree avoid overfitting. Tuning the value of `cost_complexity` helps by [pruning](https://bradleyboehmke.github.io/HOML/DT.html#pruning) back our tree. It adds a cost, or penalty, to error rates of more complex trees; a cost closer to zero is more likely to result in an overfit tree, but a larger cost can result in the opposite problem- underfitting. Tuning `tree_depth`, on the other hand, helps by [stopping our tree from growing](https://bradleyboehmke.github.io/HOML/DT.html#early-stopping) after it reaches a certain depth. We want to tune these hyperparameters to find out what those two values should be to help our model predict image segmentation better.
```
Comment (juliasilge) [inline content/start/tuning/index.Rmarkdown]:
```suggestion
We will tune the model hyperparameters to avoid overfitting. Tuning the value of `cost_complexity` helps by [pruning](https://bradleyboehmke.github.io/HOML/DT.html#pruning) back our tree. It adds a cost, or penalty, to error rates of more complex trees; a cost closer to zero is more likely to result in an overfit tree, but a larger cost can result in the opposite problem, underfitting. Tuning `tree_depth`, on the other hand, helps by [stopping our tree from growing](https://bradleyboehmke.github.io/HOML/DT.html#early-stopping) after it reaches a certain depth. We want to tune these hyperparameters to find what those two values should be for our model to do the best job predicting image segmentation. 
```

Slight edits for clarity, plus less repetition of "help"

File: content/start/tuning/index.Rmarkdown [hunk 3]
```diff
@@ -66,23 +88,48 @@ tune_spec <-
 tune_spec
 ```
 
-We can't train this specification on a single data set (such as the entire training set) and learn what the hyperparameter values should be, but we can train many models using resampled data and see which models turn out best. We can create a regular grid of values to try using some convenience functions for each hyperparameter:
+Think of `tune()` here as a placeholder. After the tuning process, we will select a single numeric value for each of these hyperparameters. But for now, we specify our parsnip model object with these `tune()` tags.
```
Comment (juliasilge) [inline content/start/tuning/index.Rmarkdown]:
```suggestion
Think of `tune()` here as a placeholder. After the tuning process, we will select a single numeric value for each of these hyperparameters. For now, we specify our parsnip model object and identify the hyperparameters we will `tune()`.
```

File: content/start/tuning/index.Rmarkdown [hunk 4]
```diff
@@ -66,23 +88,48 @@ tune_spec <-
...
+Armed with our grid filled with 25 candidate decision tree models, let's create [cross-validation folds](/start/resampling/) for tuning:
 
 ```{r cell-folds, dependson="cell-split"}
 set.seed(234)
 cell_folds <- vfold_cv(cell_train)
-
 ```
 
-Then let's use `tune_grid()` to fit models at all the different values we chose for each tuned hyperparameter. There are several options for building the object for tuning. You can tune a model specification along with a recipe or model, or you can tune a [`workflow()`](https://tidymodels.github.io/workflows/) that bundles together a model specification and a recipe or model preprocessor. Here we use a `workflow()` with a straightforward formula; if this model required more involved data preprocessing, we could use `add_recipe()` instead of `add_formula()`.
+Tuning in tidymodels will require a resampled object created with the [rsample](https://tidymodels.github.io/rsample/) package.
```
Comment (juliasilge) [inline content/start/tuning/index.Rmarkdown]:
```suggestion
Tuning in tidymodels requires a resampled object created with the [rsample](https://tidymodels.github.io/rsample/) package.
```

File: content/start/tuning/index.Rmarkdown [hunk 5]
```diff
@@ -111,56 +166,91 @@ tree_res %>%
   geom_point() +
   facet_wrap(~ .metric, scales = "free", nrow = 2) +
   scale_x_log10()
+```
+
+We can see that our "stubbiest" tree, with a depth of 1, is the worst model according to both metrics and across all candidate values of `cost_complexity`. Our deepest tree tested, with a depth of 15, did better. But, the best tree seems to be between these values with a tree depth of 4. The [`show_best()`](https://tidymodels.github.io/tune/reference/show_best.html) function shows us the top 5 candidate models by default:
```
Comment (juliasilge) [inline content/start/tuning/index.Rmarkdown]:
```suggestion
We can see that our "stubbiest" tree, with a depth of 1, is the worst model according to both metrics and across all candidate values of `cost_complexity`. Our deepest tree, with a depth of 15, did better. However, the best tree seems to be between these values with a tree depth of 4. The [`show_best()`](https://tidymodels.github.io/tune/reference/show_best.html) function shows us the top 5 candidate models by default:
```

File: content/start/tuning/index.Rmarkdown [hunk 6]
```diff
@@ -91,16 +138,24 @@ tree_wf <- workflow() %>%
...
 
-Once we have our tuning results, we can both explore them through visualization and then select the best result.
+Once we have our tuning results, we can both explore them through visualization and then select the best result. The function `collect_metrics()` gives us a tidy tibble with all the results. We had 25 candidate models and two metrics, `accuracy` and `roc_auc`, and we get a row for each `.metric` and model. 
+
+```{r collect-trees, dependson="tree-res"}
+tree_res %>% 
+  collect_metrics()
+```
+
+But we might get more out of plotting these results:
```
Comment (juliasilge) [inline content/start/tuning/index.Rmarkdown]:
```suggestion
We might get more out of plotting these results:
```

File: content/start/tuning/index.Rmarkdown [hunk 7]
```diff
@@ -111,56 +166,91 @@ tree_res %>%
...
+  pull_workflow_fit() %>% 
+  vip()
 ```
 
 These are the automated image analysis measurements that are the most important in driving segmentation quality predictions.
 
-Finally, let's return to our test data and estimate the model performance we expect to see on new data. We can use the function `last_fit()` with our finalized model; this function _fits_ the finalized model on the training data and _evaluates_ the finalized model on the testing data.
+### The last fit
+
+Finally, let's return to our test data and estimate the model performance we expect to see on new data. We can use the function [`last_fit()`](https://tidymodels.github.io/tune/reference/last_fit.html) with our finalized model; this function _fits_ the finalized model on the full training data set and _evaluates_ the finalized model on the testing data.
```
Comment (juliasilge) [inline content/start/tuning/index.Rmarkdown]:
```suggestion
Finally, let's return to our test data and estimate the model performance we expect to see with new data. We can use the function [`last_fit()`](https://tidymodels.github.io/tune/reference/last_fit.html) with our finalized model; this function _fits_ the finalized model on the full training data set and _evaluates_ the finalized model on the testing data.
```

**Review by juliasilge — APPROVED** [review]
This looks excellent; really great edits! 🙌 

@topepo I think you should especially read the very last paragraph to check the new text about tuning `min_n` and see if you have any suggested changes.

File: content/start/tuning/index.Rmarkdown (hunk 1 above)
Comment (topepo) [reply content/start/tuning/index.Rmarkdown]: I'd edit Julia's suggestion to change

> that they over-learn patterns present in the training data that do not help when faced with new data to predict. 

to 

> that they over-learn patterns present in the training data that end up being _detrimental_ when predicting new data.

File: content/start/tuning/index.Rmarkdown (hunk 5 above)
Comment (topepo) [reply content/start/tuning/index.Rmarkdown]: Good but just fyi, in the literature, a tree with a single split is a "stump". I'm not making this up.

File: content/start/tuning/index.Rmarkdown [hunk 8]
```diff
@@ -13,46 +13,68 @@ source(here::here("content/start/common.R"))
...
+`r req_pkgs(pkgs)`
+
+```{r eval=FALSE}
+library(tidymodels)  # for the tune package, along with the rest of tidymodels
+
+# Helper packages
+library(modeldata)   # for the cells data
+library(vip)         # for variable importance plots
+
+# Modeling packages
```
Comment (topepo) [inline content/start/tuning/index.Rmarkdown]: Can we remove these lines loading `rpart`?

File: content/start/tuning/index.Rmarkdown (hunk 1 above)
Comment (apreshill) [reply content/start/tuning/index.Rmarkdown]: I'll edit that locally then push, thank you

**Comment by apreshill** [comment]
@juliasilge are you ok if I change the plot? I'd like to make tree depth an ordered factor to activate the ggplot2 default of using viridis palettes for ordered factors. I (a) just like the colors better 💐 and (b) believe they are more colorblind-friendly. The only thing this would change is the colors.

**Comment by apreshill** [comment]
Please let me know if this code/plot works for you- I passed it through a colorblind image viewer and it was better (not perfect) compared to the default palette: https://deploy-preview-121--tidymodels-org.netlify.app/start/tuning/#tune-grid
